src/game/Server/WorldSession.cpp

```diff
@@ -1300,3 +1301,33 @@ void WorldSession::HandleWardenDataOpcode(WorldPacket& recv_data)
 {
     m_anticheat->WardenPacket(recv_data);
 }
+
+void WorldSession::HandleTC9PrepareForRedirect(WorldPacket& /*recvData*/)
+{
+    if (!sToCloud9Sidecar->ClusterModeEnabled())
+        return;
+
+    Player * player = this->GetPlayer();
+    if (player == nullptr)
+    {
+        WorldPacket data(TC9_SMSG_READY_FOR_REDIRECT, 1);
+        data << uint8(1); // 1 - Failed.
+        SendPacket(data);
+    }
+
+    DEBUG_LOG("Starting saving, AccountId = %d", GetAccountId());
+
+    player->SaveToDB();
+
+    // TODO: unfortunately mangos doesn't provide API to track saving progress :(
```

Comment:
We have async db queries, we can *probably* make some changes to allow for SaveToDB to run a callback on completion

Reply:
Yeah, it would be nice, but at the moment there is workaround - https://github.com/walkline/ToCloud9/commit/fd906b5e58ae735523e99f64dc016965e4e777a8#diff-cf09120953fd248dcecc81b97b5a684f7ecba5ec076f31f6cf776dac99c9b27cR130